atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -1637,6 +1649,189 @@ public void validateLocksOnCommitIfEmptyReadsFollowedByNonEmptyReadsIfValidation
+
+        verify(spiedTimeLockService, times(1)).refreshLockLeases(ImmutableSet.of(res.getLock()));
+    }
+
+    @Test
+    public void throwsIfReadingMoreThanExpectedNumberOfValues() {
+        putCellsInTable(List.of(TEST_CELL, TEST_CELL_2), TABLE_SWEPT_THOROUGH);
+
+        TimelockService spiedTimeLockService = spy(timelockService);
+        long transactionTs = spiedTimeLockService.getFreshTimestamp();
+        LockImmutableTimestampResponse res = spiedTimeLockService.lockImmutableTimestamp();
+
+        Transaction transaction = getSnapshotTransactionWith(
+                spiedTimeLockService, () -> transactionTs, res, PreCommitConditions.NO_OP, true);
+
+        assertThatThrownBy(() -> transaction.getWithExpectedNumberOfCells(
+                        TABLE_SWEPT_THOROUGH, ImmutableSet.of(TEST_CELL, TEST_CELL_2), 1))
+                .isInstanceOf(SafeIllegalStateException.class)
+                .hasMessageContaining("KeyValueService returned more results than Get expected");
+
+        verify(spiedTimeLockService, never()).refreshLockLeases(any());
+    }
+
+    @Test
+    public void keepNumberOfExpectedCellsTheSameIfNoneCached() {
+        putCellsInTable(List.of(TEST_CELL, TEST_CELL_2, TEST_CELL_4), TABLE_SWEPT_THOROUGH);
+
+        TimelockService spiedTimeLockService = spy(timelockService);
+        long transactionTs = spiedTimeLockService.getFreshTimestamp();
+        LockImmutableTimestampResponse res = spiedTimeLockService.lockImmutableTimestamp();
+
+        CacheMetrics metrics = mock(CacheMetrics.class);
+        TransactionScopedCache emptyCache = TransactionScopedCacheImpl.create(
+                ValueCacheSnapshotImpl.of(
+                        HashMap.empty(), HashSet.of(TABLE_SWEPT_THOROUGH), ImmutableSet.of(TABLE_SWEPT_THOROUGH)),
+                metrics);
+        LockWatchManagerInternal mockLockWatchManager = mock(LockWatchManagerInternal.class);
+        when(mockLockWatchManager.getTransactionScopedCache(anyLong())).thenReturn(emptyCache);
+
+        PathTypeTracker pathTypeTracker = PathTypeTrackers.constructSynchronousTracker();
+        SnapshotTransaction spiedSnapshotTransaction =
+                spy(getSnapshotTransactionWith(transactionTs, res, mockLockWatchManager, pathTypeTracker));
+        Transaction transaction = transactionWrapper.apply(spiedSnapshotTransaction, pathTypeTracker);
+
+        // Fetching 3 cells, but expect only 2 to be present, for example
+        assertThatCode(() -> transaction.getWithExpectedNumberOfCells(
+                        TABLE_SWEPT_THOROUGH, ImmutableSet.of(TEST_CELL, TEST_CELL_2, TEST_CELL_3), 2))
+                .doesNotThrowAnyException();
+
+        // We shouldn't check for locks even though we haven't fetched all 3 cells, because we fetched 2 and passed
+        // that as the expected value
+        verify(spiedTimeLockService, never()).refreshLockLeases(any());
+
+        // Since one cell is cached, should only expect 1 to be present on internal call
```

Comment:
I think this comment is no longer intended

Reply:
Oops, good catch!